### PR TITLE
Mode and difficulty switch in 12 games

### PR DESCRIPTION
**Pull request by mcmachado**
In order to submit small PRs I'm changing just 12 games for now. I'll keep working on the others. A summary of the current status is below. I won't deal with 2-player modes.

- **Games in which mode and difficulty switch are already implemented:**
`Alien`, `Amidar`, `BeamRider`, `ChopperCommand`, `DemonAttack`, `Freeway`, `Frostbite`, `Jamesbond`, `JourneyEscape`, `QBert`, `RiverRaid`, `Seaquest`.

- **Games in which mode and difficulty switch are not implemented yet:**
`Adventure`, `AirRaid`, `Asteroids`, `Atlantis`, `BankHeist`, `BattleZone`, `Berzerk`, `Bowling`, `Boxing`, `Breakout`, `Centipede`, `CrazyClimber`, `Defender`, `DoubleDunk`, `FishingDerby`, `Frogger`, `Galaxian`, `Gopher`, `Gravitar`, `Hero`, `IceHockey`, `Kaboom`, `Kangaroo`, `KingKong`, `Koolaid`, `LostLuggage`, `MrDo`, `MsPacMan`, `NameThisGame`, `Pong`, `Pooyan`, `PrivateEye`, `SirLancelot`, `SpaceInvaders`, `StarGunner`, `Tennis`, `TimePilot`, `Trondead`, `Turmoil`, `Tutankham`, `UpNDown`, `Venture`, `VideoPinball`, `WizardOfWor`, `YarsRevenge`.

- **Games that have only one one-player mode and difficulty:**
`Assault`, `Asterix`, `Carnival`, `ElevatorAction`, `Enduro`, `KungFuMaster`, `MontezumaRevenge`, `Phoenix`, `Pitfall`, `RoadRunner`, `Robotank`, `DonkeyKong`, `KeystoneKapers`, `LaserGates`, `Solaris`, `Tetris`.

- **Games that the only modes I'm not currently supporting are two-player modes**:
`Asterix`, `BeamRider`, `Carnival`, `ChopperCommand`, `DemonAttack`, `Frostbite`, `JourneyEscape`, `KungFuMaster`, `RiverRaid`, `RoadRunner`, `Seaquest`, `WizardOfWor`, `Zaxxon`.

### CODE REVIEW

File: src/games/supported/Alien.cpp [hunk 1]
```diff
@@ -142,3 +142,41 @@ void AlienSettings::loadState(Deserializer & ser) {
...
+// the given mode must be one returned by the previous function
+void AlienSettings::setMode(game_mode_t m, System &system,
+                              std::unique_ptr<StellaEnvironmentWrapper> environment) {
+
+    if(m < getNumModes()) {
+        // read the mode we are currently in
+        unsigned char mode = readRam(&system, 0x81);
+        // press select until the correct mode is reached
+        while (mode != m) {
+            environment->pressSelect(1);
```
Comment (mgbellemare) [inline src/games/supported/Alien.cpp]: Consider using the default `pressSelect()`.

File: src/games/supported/BeamRider.hpp [hunk 2]
```diff
@@ -68,6 +68,12 @@ class BeamRiderSettings : public RomSettings {
 
         virtual int lives() { return isTerminal() ? 0 : m_lives; }
 
+        // returns a list of difficulties that the game can be played in
+        // in this game, there are 2 available difficulties. The game's
+        // manual says there are three modes but at least from my ROM
+        // I couldn't find them. I found two difficulties.
+        DifficultyVect getAvailableDifficulties();
```
Comment (mgbellemare) [inline src/games/supported/BeamRider.hpp]: I think the Atari 2600 only has two levels of difficulty: Advanced & Basic. Can you update the comment?

File: src/games/supported/ChopperCommand.cpp [hunk 3]
```diff
@@ -129,3 +131,36 @@ void ChopperCommandSettings::loadState(Deserializer & ser) {
...
+    ModeVect modes = {0, 2};
+    return modes;
+}
+
+// set the mode of the game
+// the given mode must be one returned by the previous function
+void ChopperCommandSettings::setMode(game_mode_t m, System &system,
+                              std::unique_ptr<StellaEnvironmentWrapper> environment) {
+
+    if(m == 0 || m == 2){
```
Comment (mgbellemare) [inline src/games/supported/ChopperCommand.cpp]: Space between ) and {.

File: src/games/supported/DemonAttack.cpp [hunk 4]
```diff
@@ -121,3 +124,37 @@ void DemonAttackSettings::loadState(Deserializer & ser) {
...
+// set the mode of the game
+// the given mode must be one returned by the previous function
+void DemonAttackSettings::setMode(game_mode_t m, System &system,
+                              std::unique_ptr<StellaEnvironmentWrapper> environment) {
+    if(m == 1 || m == 3 || m == 5 || m == 7) {
+        // read the mode we are currently in
+        unsigned char mode = readRam(&system, 0xEA);
+        // press select until the correct mode is reached
+        while (mode != m) {
+            m_level_change = true;
```
Comment (mgbellemare) [inline src/games/supported/DemonAttack.cpp]: This seems unnecessary. Doesn't it get set below?

File: src/games/supported/JamesBond.cpp [hunk 5]
```diff
@@ -134,3 +134,34 @@ void JamesBondSettings::loadState(Deserializer & ser) {
...
+// set the mode of the game
+// the given mode must be one returned by the previous function
+void JamesBondSettings::setMode(game_mode_t m, System &system,
+                              std::unique_ptr<StellaEnvironmentWrapper> environment) {
+
+    if(m == 0 || m == 1) {
+        // read the mode we are currently in
+        unsigned char mode = readRam(&system, 0x8C);
+        // press select until the correct mode is reached
+        //in the welcome screen, the value of the mode is increased by 0x48
```
Comment (mgbellemare) [inline src/games/supported/JamesBond.cpp]: Space between // and i.

File: src/games/supported/Zaxxon.cpp [hunk 6]
```diff
@@ -132,3 +132,25 @@ void ZaxxonSettings::loadState(Deserializer & ser) {
...
+// set the mode of the game
+// the given mode must be one returned by the previous function
+void ZaxxonSettings::setMode(game_mode_t m, System &system,
+                              std::unique_ptr<StellaEnvironmentWrapper> environment) {
+
+    // read the mode we are currently in
+    unsigned char mode = readRam(&system, 0x82);
+    // press select until the correct mode is reached
+    while (mode != m) {
+        environment->pressSelect(10);
```
Comment (mgbellemare) [inline src/games/supported/Zaxxon.cpp]: Worth commenting on '10'.

File: src/games/supported/Alien.cpp (hunk 1 above)
Comment (mcmachado) [reply src/games/supported/Alien.cpp]: Done.

File: src/games/supported/ChopperCommand.cpp (hunk 3 above)
Comment (mcmachado) [reply src/games/supported/ChopperCommand.cpp]: Done.

File: src/games/supported/BeamRider.hpp (hunk 2 above)
Comment (mcmachado) [reply src/games/supported/BeamRider.hpp]: I misread something in its manual. I removed the comment. However, it is still weird. The manual reads:

```
4. Select level of difficulty by pressing it's corresponding number on the
keypad.

level: Begins at sector:
1 1
2 5
3 10

You will immediately be in the Space Station. It's control panel
displays your score, two reserve ships, sector, number of enemy saucers
in sector and three torpedoes.
```

However, nor even in the Stella I was able to change modes (and it is not being supported in the previous PR). For now I'll leave it with 2 difficulties and zero modes.

File: src/games/supported/DemonAttack.cpp (hunk 4 above)
Comment (mcmachado) [reply src/games/supported/DemonAttack.cpp]: Done.

File: src/games/supported/JamesBond.cpp (hunk 5 above)
Comment (mcmachado) [reply src/games/supported/JamesBond.cpp]: Good catch!

File: src/games/supported/Zaxxon.cpp (hunk 6 above)
Comment (mcmachado) [reply src/games/supported/Zaxxon.cpp]: I'm not sure what to say here. Each game requires an arbitrary number of presses to the select button. This one is `10`, James Bond is `20`, Freeway is `1`.

Comment (mgbellemare) [reply src/games/supported/Zaxxon.cpp]: Just a comment to indicate this isn't a typo. "Hold select button for 10 frames".

Comment (mcmachado) [reply src/games/supported/Zaxxon.cpp]: Ah, ok. Sure. I'll merge it now.

**Pull request merged**